radix-engine/src/track/db_key_mapper.rs

```diff
@@ -1,71 +1,147 @@
 };
 use sbor::rust::prelude::*;
+use utils::{combine, copy_u8_array};
 
 /// A mapper between the business RE Node / Module / Substate IDs and database keys.
 pub trait DatabaseKeyMapper {
     /// Converts the given RE Node and Module ID to the database partition's key.
+    /// Note: contrary to the sort key, we do not provide the inverse mapping here (i.e. if you
+    /// find yourself needing to map the database partition key back to RE Node and Module ID, then
+    /// you are most likely using the "partition vs sort key" construct in a wrong way).
     fn to_db_partition_key(node_id: &NodeId, module_id: ModuleId) -> DbPartitionKey;
 
     /// Converts the given Substate key to the database's sort key.
     fn to_db_sort_key(key: &SubstateKey) -> DbSortKey;
+
+    /// Converts the given database's sort key to a Substate key.
+    fn from_db_sort_key(db_sort_key: &DbSortKey) -> SubstateKey;
 }
 
 /// A [`DatabaseKeyMapper`] tailored for databases which cannot tolerate long common prefixes
 /// among keys (for performance reasons). In other words, it spreads the keys "evenly" (i.e.
-/// pseudo-randomly) across the key space.
-/// For context: our actual use-case for it is the Jellyfish Merkle Tree.
+/// pseudo-randomly) across the key space. For context: our use-case for this is the Jellyfish
+/// Merkle Tree.
+///
+/// This implementation is the actual, protocol-enforced one, to be used in public Radix networks.
+///
+/// This implementation achieves the prefix-spreading by:
+/// - using a (long, ~100% unique) hash instead of plain RE Node and Module ID (please note that it
+///   makes this mapping effectively irreversible);
+/// - using a (shorter, but hard to crack) hash prefix for Substate key.
 pub struct SpreadPrefixKeyMapper;
 
 impl DatabaseKeyMapper for SpreadPrefixKeyMapper {
     fn to_db_partition_key(node_id: &NodeId, module_id: ModuleId) -> DbPartitionKey {
         let mut buffer = Vec::new();
         buffer.extend(node_id.as_ref());
         buffer.push(module_id.0);
-        let hash_bytes = hash(buffer).0[(Hash::LENGTH - 26)..Hash::LENGTH].to_vec(); // 26 bytes
+        let hash_bytes = hash(buffer).0[..Self::PARTITION_KEY_HASH_LENGTH].to_vec();
         DbPartitionKey(hash_bytes)
     }
 
     fn to_db_sort_key(key: &SubstateKey) -> DbSortKey {
-        let bytes = match key {
-            SubstateKey::Tuple(field) => {
-                vec![*field]
-            }
-            SubstateKey::Map(key) => {
-                hash(key).0[12..Hash::LENGTH].to_vec() // 20 bytes
-            }
-            SubstateKey::Sorted((bucket, key)) => {
-                let mut bytes = bucket.to_be_bytes().to_vec();
-                bytes.extend(hash(key).0[12..Hash::LENGTH].to_vec()); // 20 bytes
-                bytes
-            }
+        let (discriminator_byte, rest_bytes) = match key {
+            SubstateKey::Tuple(field) => (Self::TUPLE_DISCRIMINATOR, vec![*field]),
```

Comment:
ooo, wait a moment, was not aware of this needed discriminator key. Let's have a quick chat on this.

Reply:
maybe it's not actually needed? see https://github.com/radixdlt/radixdlt-scrypto/pull/989#discussion_r1181718054

Reply:
I mean: it is NOT needed if all places that call `list()` know what flavor of substate key they expect.
is it a valid assumption? (e.g. in this line https://github.com/radixdlt/radixdlt-scrypto/blob/2b66d33162f4e132d1199762b7cd90d360246c31/radix-engine-queries/src/query/traverse.rs#L181)

Reply:
Yes, I think we can just remove this as part of the agreement with a client is that the client will only use a single SubstateKey type for a given partition. Otherwise, the result is undefined.

Reply:
I tried to apply the "client will only use a single SubstateKey type for a given partition" rule (see the latest commit) and there are just 2 cases where it is not obvious; hopefully I will understand more after our call 👍

Reply:
If we agreed to add the length to the RocksDB key to allow separating it into parts, I feel we perhaps should eat an extra byte here (at the end?) to allow easy decoding?

Reply:
> a typeinfo lookup

o_O do you mean like "another db read"? is it THAT bad?
I thought that we have (or easily can have) a helper like `resolveSubstateKeyType(node_id, module_id)`...?

Reply:
In the Core API we're mapping historic transaction receipts where we have no way of reading historic state.

Reply:
> I disagree with removing the discriminator UNLESS we make the key kind statically derivable from the DbPartitionKey.

To an untrained eye (i.e. mine), it looks like we can statically derive this from `NodeId.entity_type() + ModuleId` (see `decode_substate_key()` in the latest commit).
⚠️ I may be wrong - I copy-pasted from [to_typed_substate_key()](https://github.com/radixdlt/radixdlt-scrypto/blob/cf1ee4170d244acfed74b855897ad60ba0a4b993/radix-engine-queries/src/typed_substate_layout.rs#L119) and filled in the gaps with wild guesses).

But if that mapping is right, and we commit to preserving this property in future, then I am actually FOR dropping the discriminator byte from substate key DB encoding. Would you be able to verify this, @talekhinezh ?

Reply:
huh, so looking at the tests, clearly my >2 pages of static substateKey type resolution is imperfect :D
the question is - is it fixable?

Reply:
You're right actually - I had this conversation with Josh last night in fact, and made it a requirement.

At least for now, the Entity Id + Module ID (soon to be Partition Number - see REP-60) will tell us the PartitionKey. This won't necessarily be true in future if we allow scrypto modules to define their partition - instead it will need a schema lookup from a substate. That's probably okay though :)

Reply:
> this mapping should be statically decodable.

even if it is possible, it is quite complex, and (most importantly) it reaches quite deep into business knowledge about node/module/substate key types. I am not sure that a database mapping layer should really have it.